prefetcher/EIP/EIP.cc

```diff
@@ -0,0 +1,871 @@
+////////////////////////////////////////////////////////////////////////
+//
+//  Code submitted for the First Instruction Prefetching Championship
+//
+//  Authors: Alberto Ros (aros@ditec.um.es)
+//           Alexandra Jimborean (alexandra.jimborean@um.es)
+//
+//  Paper #30: The Entangling Instruction Prefetcher
+//
+////////////////////////////////////////////////////////////////////////
+
+#include "ooo_cpu.h"
+
+#define L1I_PQ_SIZE 32
+#define L1I_MSHR_SIZE 8
+#define L1I_SET 64
+#define L1I_WAY 8
+
+#define DEBUG(x) 
+
+// To access cpu in my functions
+uint32_t l1i_cpu_id;
+
+uint64_t l1i_last_basic_block;
+uint32_t l1i_consecutive_count;
+uint32_t l1i_basic_block_merge_diff;
+
+// LINE AND MERGE BASIC BLOCK SIZE
+
+#define L1I_MERGE_BBSIZE_BITS 7
+#define L1I_MERGE_BBSIZE_MAX_VALUE ((1 << L1I_MERGE_BBSIZE_BITS) - 1)
+
+// TIME AND OVERFLOWS
+
+#define L1I_TIME_DIFF_BITS 20
+#define L1I_TIME_DIFF_OVERFLOW ((uint64_t)1 << L1I_TIME_DIFF_BITS)
+#define L1I_TIME_DIFF_MASK (L1I_TIME_DIFF_OVERFLOW - 1)
+
+#define L1I_TIME_BITS 12
+#define L1I_TIME_OVERFLOW ((uint64_t)1 << L1I_TIME_BITS)
+#define L1I_TIME_MASK (L1I_TIME_OVERFLOW - 1)
+
+uint64_t l1i_get_latency(uint64_t cycle, uint64_t cycle_prev)
+{
+  uint64_t cycle_masked = cycle & L1I_TIME_MASK;
+  uint64_t cycle_prev_masked = cycle_prev & L1I_TIME_MASK;
+  if (cycle_prev_masked > cycle_masked) {
+    return (cycle_masked + L1I_TIME_OVERFLOW) - cycle_prev_masked;
+  }
+  return cycle_masked - cycle_prev_masked;
+}
+
+// ENTANGLED COMPRESSION FORMAT
+
+#define L1I_ENTANGLED_MAX_FORMATS 7
+
+// HISTORY TABLE
+
+#define L1I_HIST_TABLE_ENTRIES 1072
+#define L1I_HIST_TABLE_MASK (L1I_HIST_TABLE_ENTRIES - 1)
+#define L1I_BB_MERGE_ENTRIES 4
+#define L1I_HIST_TAG_BITS 58
+#define L1I_HIST_TAG_MASK (((uint64_t)1 << L1I_HIST_TAG_BITS) - 1)
+
+typedef struct __l1i_hist_entry {
+  uint64_t tag;       // L1I_HIST_TAG_BITS bits
+  uint64_t time_diff; // L1I_TIME_DIFF_BITS bits
+  uint32_t bb_size;   // L1I_MERGE_BBSIZE_BITS bits
+} l1i_hist_entry;
+
+l1i_hist_entry l1i_hist_table[NUM_CPUS][L1I_HIST_TABLE_ENTRIES];
+uint64_t l1i_hist_table_head[NUM_CPUS];      // log_2 (L1I_HIST_TABLE_ENTRIES)
+uint64_t l1i_hist_table_head_time[NUM_CPUS]; // 64 bits
+
+void l1i_init_hist_table()
+{
+  l1i_hist_table_head[l1i_cpu_id] = 0;
+  l1i_hist_table_head_time[l1i_cpu_id] = current_core_cycle[l1i_cpu_id];
+  for (uint32_t i = 0; i < L1I_HIST_TABLE_ENTRIES; i++) {
+    l1i_hist_table[l1i_cpu_id][i].tag = 0;
+    l1i_hist_table[l1i_cpu_id][i].time_diff = 0;
+    l1i_hist_table[l1i_cpu_id][i].bb_size = 0;
+  }
+}
+
+uint64_t l1i_find_hist_entry(uint64_t line_addr)
+{
+  uint64_t tag = line_addr & L1I_HIST_TAG_MASK;
+  for (uint32_t count = 0, i = (l1i_hist_table_head[l1i_cpu_id] + L1I_HIST_TABLE_MASK) % L1I_HIST_TABLE_ENTRIES; count < L1I_HIST_TABLE_ENTRIES;
+       count++, i = (i + L1I_HIST_TABLE_MASK) % L1I_HIST_TABLE_ENTRIES) {
+    if (l1i_hist_table[l1i_cpu_id][i].tag == tag)
+      return i;
+  }
```

Comment:
The iteration of `i` here is really strange and difficult to parse. It looks like iterating from the head backwards to the beginning, then from the end back to the head. This seems very strange. Would it be possible to use two `std::find_if`s?

```suggestion
  auto tag_check = [tag](const auto& x){ return x.tag == tag; };
  std::reverse_iterator head_rit{std::next(std::begin(l1i_hist_table[l1i_cpu_id]), l1i_hist_table_head[l1i_cpu_id])};
  auto found = std::find_if(head_rit, std::rend(l1i_hist_table[l1i_cpu_id]), tag_check);
  if (found == std::rend(l1i_hist_table[l1i_cpu_id]))
    found = std::find_if(std::rbegin(l1i_hist_table[l1i_cpu_id]), head_rit, tag_check);
  if (found == std::rend(l1i_hist_table[l1i_cpu_id]))
    return L1I_HIST_TABLE_ENTRIES;
  return std::distance(std::begin(l1i_hist_table[l1i_cpu_id]), found.base());
```